docs/proposals/push-based-metrics-collection.md

```diff
@@ -0,0 +1,77 @@
+
+In the procedure of tuning hyperparameters, Metrics Collector, which is implemented as a sidecar container attached to each training container in the [current design](https://github.com/kubeflow/katib/blob/master/docs/proposals/metrics-collector.md), will collect training logs from Trials once the training is complete. Then, the Metrics Collector will parse training logs to get appropriate metrics like accuracy or loss and pass the evaluation results to the HyperParameter tuning algorithm.
+
+However, current implementation of Metrics Collector is pull-based, raising some [design problems](https://github.com/kubeflow/training-operator/issues/722#issuecomment-405669269) such as determining the frequency we scrape the metrics, performance issues like the overhead caused by too many sidecar containers, and restrictions on developing environments which must support sidecar containers. Thus, we should implement a new API for Katib Python SDK to offer users a push-based way to store metrics directly into the Kaitb DB and resolve those issues raised by pull-based metrics collection.
+
+![](../images/push-based-metrics-collection.png)
+Fig.1 Architecture of the new design
+
+## Goal
+1. **A new parameter in Python SDK function `tune`**: allow users to specify the method of collecting metrics(push-based/pull-based).
+2. **A code injection function in mutating webhook**: recognize the metrics output lines and replace them with push-based metrics collection code.
+3. The final metrics of worker pods should be **pushed to Katib DB directly** in the push mode of metrics collection.
+
+## API
+
+### New Parameter in Python SDK Function `tune`
+
+We decided to add `metrics_collection_mechanism` to `tune` function in Python SDK.
+
+```Python
+def tune(
+    self,
+    name: str,
+    objective: Callable,
+    parameters: Dict[str, Any],
+    base_image: str = constants.BASE_IMAGE_TENSORFLOW,
+    namespace: Optional[str] = None,
+    env_per_trial: Optional[Union[Dict[str, str], List[Union[client.V1EnvVar, client.V1EnvFromSource]]]] = None,
+    algorithm_name: str = "random",
+    algorithm_settings: Union[dict, List[models.V1beta1AlgorithmSetting], None] = None,
+    objective_metric_name: str = None,
+    additional_metric_names: List[str] = [],
+    objective_type: str = "maximize",
+    objective_goal: float = None,
+    max_trial_count: int = None,
+    parallel_trial_count: int = None,
+    max_failed_trial_count: int = None,
+    resources_per_trial: Union[dict, client.V1ResourceRequirements, None] = None,
+    retain_trials: bool = False,
+    packages_to_install: List[str] = None,
+    pip_index_url: str = "https://pypi.org/simple",
+    metrics_collection_mechanism: str = "pull", # The newly added parameter
+)
+```
+
+## Implementation
```

Comment:
Do you mean that Trial controller should have a response like StatusCode in http from Katib DB and process it in the further steps?

Reply:
I think, we should just verify if Trial controller will update Trial observation correctly in case of [`None` Metrics Collector.](https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/trial/trial_controller_util.go#L125-L133)

Reply:
Okay, I get your idea. Do you mean that Trial controller will report an error if we do not change it to be compatible with push-based metrics collection?

Reply:
Currently, metrics collector inserts `unavailable` value for each metrics into DB if that metrics can't be found: https://github.com/kubeflow/katib/blob/master/pkg/metricscollector/v1beta1/file-metricscollector/file-metricscollector.go#L181-L193.
And since [`isObservationAvailable` returns false](https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/trials/v1beta1/util.go#L94), we marked Trial with [Metrics Unavailable status](https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/trial/trial_controller_util.go#L69). We need to see how will it work for push-based Metrics collector.

Reply:
Metrics collectors won't be injected into each trial pod when the push-based metrics collection is applied. Instead, training metrics will be pushed to Katib DB directly as users call `katib_client.report_metrics` for metrics collection. So, this issue will not happen in push-based metrics collection scenario. WDYT👀 @andreyvelich

Reply:
Yeah, I think that is right. Initially, I implemented this logic for pull-based metrics in this PR: https://github.com/kubeflow/katib/pull/1333
Since sometime Trials can be in succeeded state even when metrics collector sidecar container is still running.
Because some CRDs (e.g. Tekton Pipeline) don't wait for all sidecar to be complete.

For the push-based metrics it is not the case, since `katib_client.report_metrics()` call is part of user's training container execution.

We still need to make appropriate changes in the controller to make sure we insert `unavailable` value into Katib DB, if user doesn't report metric accidentally. What do you think @johnugeorge @tenzen-y @Electronic-Waste ?

Reply:
I agree with you. I'll investigate it.

Reply:
I noticed that the re-queue logic is implemented in the following lines:

```
// If observation is empty metrics collector doesn't finish.
// For early stopping metrics collector are reported logs before Trial status is changed to EarlyStopped.
if jobStatus.Condition == trialutil.JobSucceeded && instance.Status.Observation == nil {
	logger.Info("Trial job is succeeded but metrics are not reported, reconcile requeued")
	return errMetricsNotReported
}
```

I decide to add a if-else statement to distinguish pull-based and push-based metrics collection. In the push-based collection, the trial does not need to be requeued. Instead, we'll insert a `unavailable` value to Katib DB and change the status of trial to `MetricsUnavailable`. WDYT👀 @andreyvelich @tenzen-y @johnugeorge @gaocegege